utilities/qneg.F90

```diff
@@ -0,0 +1,388 @@
+  use shr_sys_mod,         only: shr_sys_flush
+!  use cam_history_support, only: max_fieldname_len
+!  use ppgrid,              only: pcols
+
+  implicit none
+  private
+  save
+
+  ! Public interface.
+
+  public :: qneg_init
+  public :: qneg_run !qneg3 in CAM6
+  public :: qneg_timestep_final
+  public :: qneg_final
+
+  ! Private module variables
+  logical          :: collect_stats = .false.
+  logical          :: timestep_reset = .false.
+  logical          :: qneg_initialized = .false.
+  integer          :: num_constituents = 0
+
+  real(kind_phys), parameter :: tol = 1.e-12_kind_phys
+  real(kind_phys), parameter :: worst_reset = 1.e35_kind_phys
+
+  ! Diagnostic field names
+!  integer              :: num_diag_fields = 0
+!  character(len=max_fieldname_len) :: diag_names(num_diag_fields)
+!  logical             :: cnst_out_calc = .false.
+!  logical             :: cnst_outfld(num_diag_fields) = .false.
+
+  ! Summary buffers
+  integer, parameter    :: num_bins = 24
+  character(len=CS)     :: qneg_warn_labels(num_bins) = ''
+  integer, allocatable  :: qneg_warn_num(:,:)
+  real(kind_phys), allocatable :: qneg_warn_worst(:,:)
+
+  private             :: qneg_print_summary
+!  private             :: calc_cnst_out
+  private             :: find_index
+  interface reset_stats
+     module procedure reset_stats_scalar
+     module procedure reset_stats_array
+  end interface reset_stats
+
+contains
+
+!> \section arg_table_qneg_init Argument Table
+!! \htmlinclude qneg_init.html
+  subroutine qneg_init(print_qneg_warn, number_of_constituents, qmin, errcode, errmsg)
+    !use cam_history,    only: addfld, horiz_only
+    !use constituents,   only: cnst_longname
+
+    character(len=*),  intent(in)  :: print_qneg_warn
+    integer,           intent(in)  :: number_of_constituents
+    real(kind_phys),   intent(out) :: qmin(:)
+    integer,           intent(out) :: errcode
+    character(len=512),intent(out) :: errmsg
+
+    character(len=*), parameter  :: subname = 'qneg_init'
+    integer                      :: ierr
+
+    errcode = 0
+    errmsg = ''
+
+    !Check if already initialized:
+    if (qneg_initialized) then
+       return
+    end if
+
+!    integer :: index
+
+!    do index = 1, num_constituents
+!       diag_names(index) = trim(cnst_name(index))//'_qneg'
+!       call addfld(diag_names(index), (/ 'lev' /), 'I', 'kg/kg',              &
+!            trim(cnst_longname(index))//' QNEG error (cell)')
+!       diag_names(num_constituents+index) = trim(cnst_name(index))//'_qneg_col'
+!       call addfld(diag_names(num_constituents+index), horiz_only, 'I', 'kg/kg',         &
+!            trim(cnst_longname(index))//' QNEG error (column)')
+!    end do
+!    diag_names((2*num_constituents) + 1) = 'qflux_exceeded'
+!    call addfld(diag_names((2*num_constituents) + 1), horiz_only, 'I', 'kg/m^2/s',     &
+!         'qflux excess (QNEG4)')
+
+    !Create fake qmin -> will be removed when qmin exists
+    qmin = 0._kind_phys
```

Comment:
Agreed that, for sure, it will eventually come from the host model. I think the decision was made to put the faked version in here since it doesn't exist anywhere right now and some design decisions have to be made about where it will come from.

Reply:
Added an issue - https://github.com/NCAR/CAMDEN/issues/158 - to CAMDEN